### PR TITLE
Add setcmdline()

### PR DESCRIPTION
**Pull request by Shougo**
It is useful to create plugins for command line.

### CODE REVIEW

**Comment by brammool** [comment]
I wonder what happens when calling setcmdline() while in custom completion.  Should have a test for that.
There might be other situations where changing the command line unexpectedly might cause trouble.

**Comment by Shougo** [comment]
> I wonder what happens when calling setcmdline() while in custom completion. 

What means the `custom completion`?
It is command custom completion?   To add tests, I need to understand it.

**Comment by brammool** [comment]
> > I wonder what happens when calling setcmdline() while in custom completion. 
> 
> What means the `custom completion`?
> It is command custom completion?   To add tests, I need to understand it.

For a user command custom completion can be defined with:
	-complete=custom,{func}

See ":help :command-completion-custom".
In the function used here, if you would call setcmdline(), I wonder if
something breaks (using freed memory, for example).

Perhaps a :cmap using <expr> might also cause trouble.

-- 
GALAHAD: No look, really, this isn't nescess ...
PIGLET:  We must examine you.
GALAHAD: There's nothing wrong with ... that.
                 "Monty Python and the Holy Grail" PYTHON (MONTY) PICTURES LTD

 /// Bram Moolenaar -- ***@***.*** -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by Shougo** [comment]
I have added tests.

**Comment by Shougo** [comment]
```
Failures: 
	From test_cmdline.vim:
	Found errors in Test_setcmdline():
	command line..script /home/runner/work/vim/vim/src/testdir/runtest.vim[468]..function RunTheTest[44]..Test_setcmdline line 33: Expected 'let foo=0' but got '"DoCmd January February Mars'
```

I don't know why it is failed in gvim only.
It works for me...

**Comment by Shougo** [comment]
I have changed the map character.  It is passed.

**Comment by vimpostor** [comment]
Do you have an example usecase for this function? I don't quite get where this is useful.

**Comment by bfrg** [comment]
> Do you have an example usecase for this function? I don't quite get where this is useful.

1. Custom command-line popup window (the popup callback function calls `setcmdline()` to set the command-line after user made selection and hit ENTER)
2. If someone has a user command with a command-completion function that obtains its list of candidates asynchronously the job callback function needs a way to set the command-line once the list of candidates are obtained.

File: runtime/doc/builtin.txt [hunk 1]
```diff
@@ -7896,6 +7901,16 @@ setcharsearch({dict})					*setcharsearch()*
...
 
+setcmdline({str} [, {pos}])					*setcmdline()*
+		Set the command line to {str} and set the cursor position to
+		{pos}.
+		If {pos} is omitted, the last position is used.
+		Returns FALSE when successful, TRUE when not editing the
+		command line.
+
+		Can also be used as a |method|: >
+			SetCmdline()->setcmdline()
```
Comment (bfrg) [inline runtime/doc/builtin.txt]: How about a better name for the first function? For example:
```
GetText()->setcmdline()
```

Comment (Shougo) [reply runtime/doc/builtin.txt]: OK.

**Comment by habamax** [comment]
> 1. Custom command-line popup window (the popup callback function calls `setcmdline()` to set the command-line after user made selection and hit ENTER)

I for now use `feedkeys(":Command")` to have it, I guess proper function would be better.

**Comment by brammool** [comment]
A few review comments (sorry it took a while):
- please use the same formatting as for other functions (function return type goes on a separate line)
- help says returning FALSE or TRUE, but it looks like it's returning one and zero
- use a better error than e_invalid_argument 
- a negative position is rejected without any error
- All declarations go at the start of the block (ANSI C)

**Comment by Shougo** [comment]
> help says returning FALSE or TRUE, but it looks like it's returning one and zero

yes, but it is the same behavior with `setcmdpos()`.  The documentation is wrong??

**Comment by Shougo** [comment]
> please use the same formatting as for other functions (function return type goes on a separate line)

Fixed.

>  a negative position is rejected without any error

Fixed.

> All declarations go at the start of the block (ANSI C)

Fixed.

**Comment by brammool** [comment]
> > help says returning FALSE or TRUE, but it looks like it's returning
> > one and zero
> 
> yes, but it is the same behavior with `setcmdpos()`.  The
> documentation is wrong?? 

Ah, you took it from there, and there it's confusing as well.

FALSE and TRUE are used for argument values, where zero and v:false are
accepted as FALSE and 1 or v:true are accepted as TRUE.  But when
returning a value we know what we are returning, thus we can just use
zero and one.  Othewise users may expect v:false or v:true, which
matters when using Vim9 script.

-- 
It is illegal to take more than three sips of beer at a time while standing.
		[real standing law in Texas, United States of America]

 /// Bram Moolenaar -- ***@***.*** -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by Shougo** [comment]
Thank you for the merge.

**Pull request closed without merging**